Rfc5 review response (#350)
ome/ngff: rfc/5/index.md

  A file format that natively supports spatial transformations allows researchers to apply, modify,
  or reverse transformations as needed for different analysis purposes.
  This flexibility is critical for tasks such as longitudinal studies, multi-modal imaging,
  and comparative analysis across different subjects or experimental conditions.

Toward these goals, this RFC expands the set of transformations in the OME-Zarr spec
covering many of the use cases requested in [this github issue](https://github.com/ome/ngff/issues/84).
It also adds "coordinate systems" - named sets of "axes."
Related the relationship of discrete arrays to physical coordinates
and the interpretation and motivation for axis types.

## Proposal

Below is a complete copy of the proposed changes including suggestions
from reviewers and contributors of the previously associated [github pull request](https://github.com/ome/ngff/pull/138).
The changes, if approved, shall be translated into bikeshed syntax and added to the ngff repository in a separate PR.
This PR will then comprise complete json schemas when the RFC enters the SPEC phase (see RFC1).

### "coordinateSystems" metadata

> A "coordinate system" is a collection of "axes" / dimensions with a name.

Reviewer: the definition of the JSON object should probably be preceded by some description of the abstract thing the JSON object models. Otherwise, readers might not know what the point it is. For example, here I would state that a coordinate system defines a mapping from array indices of Zarr arrays to points in a domain with some scientific or application-specific meaning, and then move to the description of the JSON object.

> Every coordinate system:

Author: I'm not sure I have a good answer to this one in the sense of a good answer that reflects a consensus in the community. This quesiton reflects a discussion in another PR (https://github.com/ome/ngff/pull/324) that eventually remained inconclusive. I would place the answer to this question outside the scope of RFC5, which doesn't mean that I think that there should be a good answer to it.

- MUST contain the field "name".

>   The value MUST be a non-empty string that is unique among all entries under `coordinateSystems`.

Author: My understanding from an implementation/viewer perspective is, that users could or would be able to select the coordinate system n which to project/view their data. Having an empty string in there would make that a bit messy, I think? Aside from that, I'd say that implementers are perfectly free to choose default values for coordinate systems to be in line so meeting this requirement is not overly burdening.

>   The value MUST be a non-empty string that is unique among all entries under `coordinateSystems`.

Reviewer: That makes sense, this reasoning should probably be part of the spec

- MUST contain the field "axes", whose value is an array of valid "axes" (see below).

The order of the `"axes"` list matters

> and defines the index of each array dimension and coordinates for points in that coordinate system.

Reviewer: ```suggestion
The elements of `"axes"` correspond to the index of each array dimension and coordinates for points in that coordinate system.
```

````{admonition} Example

Coordinate Systems metadata example

```json
{
    "name" : "volume_micrometers",
    "axes" : [
        {"name": "z", "type": "space", "unit": "micrometer"},
        {"name": "y", "type": "space", "unit": "micrometer"},
        {"name": "x", "type": "space", "unit": "micrometer"}
    ]
}
```
````

The axes of a coordinate system (see below) give information
about the types, units, and other properties of the coordinate system's dimensions.
Axis names may contain semantically meaningful information, but can be arbitrary.
As a result, two coordinate systems that have identical axes in the same order 
may not be "the same" in the sense that measurements at the same point
refer to different physical entities and therefore should not be analyzed jointly.
Tasks that require images, annotations, regions of interest, etc.,
SHOULD ensure that they are in the same coordinate system (same name and location within the Zarr hierarchy, with identical axes)
or can be transformed to the same coordinate system before doing analysis.
See the [example below](example:coordinate_transformation).

#### "axes" metadata

"axes" describes the dimensions of a coordinate systems

> and adds an interpretation to the samples along that dimension.

Reviewer: We are defining a list of objects first, and then describing the properties of the elements of the list. IMO it's more logical to define the axis object first, and then define `axes` as a list of axis objects.

It is a list of dictionaries,

> where each dictionary describes a dimension (axis) and:

Author: Again, see the discussion over at https://github.com/ome/ngff/pull/324. I'm personally in flavor of declaring all metadata under the `ome` key "off limits" for extra data or handing out namespaced keys in defined locations [as rfc8 suggests](https://github.com/ome/ngff/pull/343) but I'm hesitant to add that kind of restriction in the scope of rfc5.

- MUST contain the field "name" that gives the name for this dimension.
  The values MUST be unique across all "name" fields in the same coordinate system.
- SHOULD contain the field "type".
  It SHOULD be one of the strings "array", "space", "time", "channel", "coordinate", or "displacement"

>   but MAY take other string values for custom axis types that are not part of this specification yet.

Author: The current version of the spec states:

> SHOULD contain the field "type". It SHOULD be one of "space", "time" or "channel", but MAY take other string values for custom axis types that are not part of this specification yet.

So I'd say that other kind of axes are already allowed and rfc5 doesn't really change that behavior much. Which is reasonable since [rfc3](https://ngff.openmicroscopy.org/rfc/3/index.html) already addresses this topic in the depth that it deserves.

- MAY contain the field "discrete".
  The value MUST be a boolean,

>   and is `true` if the axis represents a discrete dimension.

Author: Small hint added in 8237047f67759fff8a4276f3669d5de54be610ca. Adding an anchor link here seemed a bit overkill because that information is following two paragraphs below:

> Arrays are inherently discrete (see Array coordinate systems, below)
but are often used to store discrete samples of a continuous variable.
The continuous values "in between" discrete samples can be retrieved using an *interpolation* method.
If an axis is continuous (`"discrete" : false`), it indicates that interpolation is well-defined.
Axes representing `space` and `time` are usually continuous.
Similarly, joint interpolation across axes is well-defined only for axes of the same `type`.
In contrast, discrete axes (`"discrete" : true`) may be indexed only by integers.
Axes representing a channel, coordinate, or displacement are usually discrete.

>   and is `true` if the axis represents a discrete dimension.

Reviewer: IMO the term should be defined when it is first introduced. It's not reasonable to expect readers of the spec to initiate a text search when they encounter an unknown term.

>   and is `true` if the axis represents a discrete dimension.

Reviewer 2: I don't see that commit 823704, but I find that a simple "(see below)" comment can be a big help. It reassures readers that they will find clarity in due course, or they can skip ahead if they need to know now.

- MAY contain the field "longName".
  The value MUST be a string,
  and can provide a longer name or description of an axis and its properties.

The length of "axes" MUST be equal to the number of dimensions of the arrays that contain the image data.

Arrays are inherently discrete (see Array coordinate systems, below)
but are often used to store discrete samples of a continuous variable.
The continuous values "in between" discrete samples can be retrieved using an *interpolation* method.
If an axis is continuous (`"discrete" : false`), it indicates that interpolation is well-defined.
Axes representing `space` and `time` are usually continuous.
Similarly, joint interpolation across axes is well-defined only for axes of the same `type`.
In contrast, discrete axes (`"discrete" : true`) may be indexed only by integers.
Axes representing a channel, coordinate, or displacement are usually discrete.

```{note}

> The most common methods for interpolation are "nearest neighbor", "linear", "cubic", and "windowed sinc".

Reviewer: do we need to list the names of popular interpolation routines? Presumably anyone using this metadata already has that level of familiarity with interpolation.

by the half-open interval `[-0.5, 0.5) x [-0.5, 0.5)` (i.e., -0.5 is included, +0.5 is excluded).
See chapter 4 and figure 4.1 of the ITK Software Guide.

### "coordinateTransformations" metadata

"coordinateTransformations" describe the mapping between two coordinate systems (defined by "coordinateSystems").
For example, to map an array's discrete coordinate system to its corresponding physical coordinates.
Coordinate transforms are in the "forward" direction.
This means they represent functions from *points* in the input space to *points* in the output space
(see [example below](example:coordinate_transformation_scale)).

They:

- MUST contain the field "type" (string).
- MUST contain any other fields required by the given "type" (see table below).
- MUST contain the field "output" (string),
  unless part of a `sequence` or `inverseOf` (see details).
- MUST contain the field "input" (string),
  unless part of a `sequence` or `inverseOf` (see details).
- MAY contain the field "name" (string).
  Its value MUST be unique across all "name" fields for coordinate transformations.
- Parameter values MUST be compatible with input and output space dimensionality (see details).

> The following transformations are supported:

Reviewer: I think this sentence needs to be translated into "specese". MUST all implementations support all and only these transformations? Is it an error to include a transformation that's _not_ in this list?

More broadly, I think this spec should define rules for adding new transformations, and omitting transforms from implementations. A few things seem true:
- people can define new, special transformations that work for their data
- implementations (like visualization tools) may only support a small set of all possible transformations
- publishing a new transformation of general interest to everyone should be a lightweight process, i.e. not an RFC.

I think this argues for a core / external split like we have in Zarr V3 w.r.t to codecs. The Zarr V3 spec defines a [handful of core codecs](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/index.html) that implementations SHOULD support. For Zarr this is not a MUST these codecs often require external libraries which not every software ecosystem has access to, but I think for ome-zarr, scale and translation don't require special dependencies and so they could be a MUST support. 

At the same time, we don't want to amend the Zarr spec every time a new codec gets popular, and we don't want every zarr implementation to worry about supporting every possible codec, so we have a lightweight mechanism for users to publish new external codecs in the [zarr-extensions repo](https://github.com/zarr-developers/zarr-extensions/tree/main/codecs). Zarr implementations have more discretion in whether they add support for these codecs.

I think a similar structure for transforms could simplify this spec. 
- define the structure of the transform object
- define that the coordinateTransformations containers take any transform object (i.e., the set of transforms is not closed)
- define some core transforms that are MUST / SHOULD support, and reserve their names to prevent confusion (we don't want a user-defined `scale` transform).
- set up an external resource for people to declare their own transforms. seed that repo with some of the more exotic transforms from this spec, e.g. the ones that have been controversial (like `inverseOf`).

Reviewer 3: There is this section below:

```markdown
Conforming readers:

- MUST parse identity, scale, translation transformations;
- SHOULD parse mapAxis, affine, rotation transformations;
- ...
```

Reviewer: I suggest replacing "parse" with "support"

Author: That's a valid point @d-v-b . I had this discussion with @bogovicj a few weeks ago and he (rightfully) pointed out that some viewers just straight up don't have the capacity (on a very fundamental level) to visualize anything beyond scale (or even translation), with Fiji as a concrete example. We came to the conclusion that "parsing" in this context would have to amount to:

- Reading the metadata
- Telling the user that the metadata cannot be displayed by the viewer

Which is what we wrote down one line below what @clbarnes quotes:

> SHOULD display an informative warning if encountering transformations that cannot be parsed;

I extended this statement to more explicitly address viewers (1687c6ebfd37f8c17a94b1e1a640e69c1d1f60e3)

I think napari does it nicely: If you display affines in 3D all is well, but if you switch to 2D, napari raises a flag to tell you that you are *not* looking at the transformed image.

    }
]
```

And the image at the path `sampleA_instrument1` would have the following as the first coordinate system:

```json
"coordinateSystems": [
    {
        "name": "sampleA-instrument1",
        "axes": [
            {"name": "z", "type": "space", "unit": "micrometer"},
            {"name": "y", "type": "space", "unit": "micrometer"},
            {"name": "x", "type": "space", "unit": "micrometer"}
        ]
    },
]
```

The image at path `sampleA_instrument2` would have this as the first listed coordinate system:

```json
[
    {
        "name": "sampleA-instrument2",
        "axes": [
            {"name": "z", "type": "space", "unit": "micrometer"},
            {"name": "y", "type": "space", "unit": "micrometer"},
            {"name": "x", "type": "space", "unit": "micrometer"}
        ]
    }
],
```
````

#### Additional details

Most coordinate transformations MUST specify their input and output coordinate systems
using `input` and `output` with a string value
that MUST correspond to the name of a coordinate system or the path to a multiscales group.
Exceptions are if the coordinate transformation appears in the `transformations` list of a `sequence`

> or is the `transformation` of an `inverseOf` transformation.

Reviewer 3: What about bijections and byDimensions?

When stored as a 2D zarr array, the first dimension indexes rows and the second dimension indexes columns
(e.g., an array of `"shape":[3,4]` has 3 rows and 4 columns).
When stored as a 2D json array, the inner array contains rows (e.g. `[[1,2,3], [4,5,6]]` has 2 rows and 3 columns).

#### Transformation types

Input and output dimensionality may be determined by the coordinate system referred to by the `input` and `output` fields, respectively. 
If the value of `input` is a path to an array, its shape gives the input dimension,
otherwise it is given by the length of `axes` for the coordinate system with the name of the `input`.
If the value of `output` is an array, its shape gives the output dimension,
otherwise it is given by the length of `axes` for the coordinate system with the name of the `output`.

##### <a name="identity">identity</a>

`identity` transformations map input coordinates to output coordinates without modification.
The position of the i-th axis of the output coordinate system
is set to the position of the ith axis of the input coordinate system.
`identity` transformations are invertible.

> The `input` and `output` fields MAY be omitted if part of a [`sequence`](#sequence) transformation.

Author: e02ba891cc81fb4766fe0432486dbccb90c10492

if those fields may be omitted for the transformation it wraps.

```{note}
Software libraries that perform image registration
often return the transformation from fixed image coordinates to moving image coordinates,
because this "inverse" transformation is most often required
when rendering the transformed moving image.
Results such as this may be enclosed in an `inverseOf` transformation.
This enables the "outer" coordinate transformation to specify the moving image coordinates
as `input` and fixed image coordinates as `output`,
a choice that many users and developers find intuitive.
```

##### <a name="sequence">sequence</a>

A `sequence` transformation consists of an ordered array of coordinate transformations,
and is invertible if every coordinate transform in the array is invertible
(though could be invertible in other cases as well).
To apply a sequence transformation to a point in the input coordinate system,
apply the first transformation in the list of transformations.
Next, apply the second transformation to the result.
Repeat until every transformation has been applied.
The output of the last transformation is the result of the sequence.

> A sequence transformation MUST NOT be part of another sequence transformation.

Reviewer 3: Is this a necessary constraint? It's obviously not the simplest way for users to express a transform, but requiring it just adds more complexity for the implementations.

Reviewer 2: For implementations that have to read this, it's simpler not to have to write a recursive function to handle unknown depth of nesting. For those that write it, it's also quite simple to avoid nesting.

Reviewer 3: Every implementation is going to define a generic "deserialise coordinate transform" routine anyway. To my mind, the difference is something like

```python
def deser_coord_transform(jso):
    match jso["type"]:
        case "sequence":
            inner = [deser_coord_transform(d) for d in jso["transformations"]]
            ...
        case ...  # all the other types
    ...
```

vs 

```python
def deser_coord_transform(jso, is_root=False):
    """is_root should only be True when called on the outermost coordinate transform"""
    match jso["type"]:
        case "sequence":
            inner = []
            for d in jso["transformations"]:
                if not is_root and d["type"] == "sequence":
                    raise BadCoordTransform()
                inner.append(deser_coord_transform(d, False))
            ...
        case ...  # all the other types
    ...
```

If you're using a serialisation framework which handles internally-tagged enums natively (pydantic, serde) then you need to either define a nearly-identical enum with all the same behaviour or add a custom post-hoc validator. There are other containers (inverseOf, bijection, byDimension) which also may each need to handle recursion differently, and stacking them in different orders looks like a headache if we make any statement except "it's fine".

It's not a lot of complexity, but every "except in the case of..." adds to the mental load of reading and writing the data and implementations.

Recursion is a problem when the bounds are unknown, but we know it's not going to get out of hand here because we've just deserialised it from a reasonably-sized JSON blob, and it's probably been written by a well-behaved writer. Strict writers (SHOULD NOT recurse), permissive readers (you can just re-use the same function).

Reviewer 3: Is the best place for discussion an issue or discussion post on this repo, or over at ngff-specs?

  the array at location path MUST have `N+1` dimensions
* The field coordinate system MUST contain an axis identical to every axis
  of its input coordinate system in the same order.
* The field coordinate system MUST contain an axis with type `coordinate` or `displacement`, respectively,
  for transformations of type `coordinates` or `displacements`.
    * This SHOULD be the last axis (contiguous on disk when c-order).
* If the output coordinate system has `M` axes,
  the length of the array along the `coordinate`/`displacement` dimension MUST be of length `M`.

The `i`th value of the array along the `coordinate` or `displacement` axis refers to the coordinate or displacement
of the `i`th output axis. See the example below.

`coordinates` and `displacements` transformations are not invertible in general,
but implementations MAY approximate their inverses.
Metadata for these coordinate transforms have the following fields: 

<dl>
  <dt><strong>path</strong></dt>
  <dd>  The location of the coordinate array in this (or another) container.</dd>
  <dt><strong>interpolation</strong></dt>
  <dd>  The <code>interpolation</code> attributes MAY be provided.

>         It's value indicates the interpolation to use

Reviewer: ```suggestion
        Its value indicates the interpolation to use
```

##### <a name="identity">identity</a>

`identity` transformations map input coordinates to output coordinates without modification.
The position of the i-th axis of the output coordinate system
is set to the position of the ith axis of the input coordinate system.
`identity` transformations are invertible.

The `input` and `output` fields MAY be omitted if part of a [`sequence`](#sequence) transformation.

##### <a name="mapAxis">mapAxis</a>

`mapAxis` transformations describe axis permutations as a transpose vector of integers.
Transformations MUST include a `mapAxis` field
whose value is an array of integers that specifies the new ordering in terms of indices of the old order.
The length of the array MUST equal the number of dimensions in both the input and output coordinate systems.
Each integer in the array MUST be a valid zero-based index into the input coordinate system's axes
(i.e., between 0 and N-1 for an N-dimensional input).
Each index MUST appear exactly once in the array.
The value at position `i` in the array indicates which input axis becomes the `i`-th output axis.
`mapAxis` transforms are invertible.

> `mapAxis` transforms are invertible.

Reviewer 4: ```suggestion
```
Duplicated line

| **Author** | Davis Bennett | @d-v-b | | 2024-07-30 | Implemented validation |
| **Author** | Luca Marconato | @LucaMarconato | EMBL | 2024-07-30 | Implemented |
| **Author** | Matt McCormick | @thewtex | ITK | 2024-07-30 | Implemented |
| **Author** | Stephan Saalfeld | @axtimwalde | HHMI Janelia | 2024-07-30 | Implemented (with JB) |
| **Author** | Johannes Soltwedel | @jo-mueller | German Bioimaging e.V. | 2025-10-07 | Implemented |
| **Endorser** | Will Moore | @will-moore | University of Dundee | 2025-10-23 | Implemented |
| **Endorser** | David Stansby | @dstansby | University College London | 2025-10-23 | Implemented |
| **Endorser** | Norman Rzepka | @normanrz | Scalable Minds | 2024-08-22 | |
| **Reviewer** | Dan Toloudis, David Feng, Forrest Collman, Nathalie GAudreault, Gideon Dunster | toloudis, dyf, fcollman | Allen Institutes | 2024-11-28 | [Review](rfcs:rfc5:review1) |
| **Reviewer** | Will Moore, Jean-Marie Burel, Jason Swedlow | will-moore, jburel, jrswedlow | University of Dundee | 2025-01-22 | [Review](rfcs:rfc5:review2)|

## Overview

This RFC provides first-class support for spatial and coordinate transformations in OME-Zarr.

> Working version title: **0.6dev2**

Reviewer 5: :+1:

The most common methods for interpolation are "nearest neighbor", "linear", "cubic", and "windowed sinc".
Here, we refer to any method that obtains values at real-valued coordinates using discrete samples as an "interpolator".
As such, label images may be interpolated using "nearest neighbor" to obtain labels at points along the continuum.
```

#### Array coordinate systems

The dimensions of an array do not have an interpretation
until they are associated with a coordinate system via a coordinate transformation.
Nevertheless, it can be useful to refer to the "raw" coordinates of the array.
Some applications might prefer to define points or regions-of-interest in "pixel coordinates" rather than "physical coordinates," for example.
Indicating that choice explicitly will be important for interoperability.
This is possible by using **array coordinate systems**.

Every array has a default coordinate system whose parameters need not be explicitly defined.
The dimensionality of each array coordinate system equals the dimensionality of its corresponding Zarr array.
Its name is the path to the array in the container,
its axes have `"type": "array"`, are unitless, and have default names.

> The i-th axis has `"name": "dim_i"` (these are the same default names used by [xarray](https://docs.xarray.dev/en/stable/user-guide/terminology.html)).

Reviewer: ```suggestion
The i-th axis has `"name": "dim_i"`.
```
I don't think it's relevant what xarray's defaults are, and also xarray might change their defaults, which would bring this section out of date. Unless xarray compatibility is an explicit design goal we should remove references to xarray from this section (fine to talk about xarray outside the normative section of the spec).

* `input` and `output` MUST have an equal number of dimensions.

##### <a name=trafo-byDimension>byDimension</a>

`byDimension` transformations build a high dimensional transformation
using lower dimensional transformations on subsets of dimensions.
The `input` and `output` fields MUST always be included for this transformations type.

<dl>
  <dt><strong>transformations</strong></dt>

>   <dd>  Each child transformation MUST contain <code>input_axes</code> and <code>output_axes</code> fields

Reviewer 3: camelCase, if that's what we're doing

in which case the `forward` transformation's `input` and `output` are understood to match the bijection's,
and the `inverse` transformation's `input` (`output`) matches the bijection's `output` (`input`),
see the example below.

The `input` and `output` fields MAY be omitted for `bijection` transformations
if the fields may be omitted for both its `forward` and `inverse` transformations

Practically, non-invertible transformations have finite extents,
so bijection transforms should only be expected to be correct / consistent for points that fall within those extents.
It may not be correct for any point of appropriate dimensionality.

### "multiscales" metadata

Metadata about an image can be found under the `multiscales` key in the group-level OME-Zarr Metadata.
Here, "image" refers to 2 to 5 dimensional data representing image
or volumetric data with optional time or channel axes.
It is stored in a multiple resolution representation.

`multiscales` contains a list of dictionaries where each entry describes a multiscale image.

Each `multiscales` dictionary MUST contain the field "coordinateSystems",

> whose value is an array containing valid coordinate system metadata

Reviewer: ```suggestion
whose value is an array containing coordinate system metadata
```

(see [coordinate systems](#coordinatesystems-metadata)).
The last entry of this array is the "intrinsic" coordinate system

> and MUST contain axis information pertaining to physical coordinates.

Author: Would it be equally well-defined if the graph is built from the top-down, though?🤔

> and MUST contain axis information pertaining to physical coordinates.

Reviewer 2: Yes. I don't see that it would be any different. When you traverse from the parent to child image and load the child image metadata, you have all the info in hand to construct the same graph. Then you can write it out or traverse it whichever way you want.

> and MUST contain axis information pertaining to physical coordinates.

Reviewer 4: It might not required for the *last* entry in the list to be special though, because transforms at the datasets level have to (or at least, in all the examples given do) reference the output coordinate system by name anyway (e.g., https://github.com/jo-mueller/ngff-rfc5-coordinate-transformation-examples/blob/ac410d33ccd3c2f537b533329421bd63ad98f613/user_stories/stitched_tiles_2d.zarr/tile_2/zarr.json#L33-L42)

> and MUST contain axis information pertaining to physical coordinates.

Reviewer 2: @dstansby in your graph implementation of RFC5, do you get a different graph based on the ordering of the `coordinateSystems` list? If so, then I imagine for my example above, with the coordinateSystems ordered `[{"name": "default"...}, {"name": "rotated"...}]` then the "default" would be included in the graph but the "rotated" coordinateSystem (and the sibling "rotation" coordinateTransform) would be completely ignored? That doesn't seem to make any sense.

> and MUST contain axis information pertaining to physical coordinates.

Reviewer 4: I would get a different graph based on the first item of the `coordinateSystems` list at the multiscales level, but otherwise graphs are identical with respect to any other orderings.

In the above, both coordinate systems would be included. I think this is the same case as the user story I added at https://github.com/jo-mueller/ngff-rfc5-coordinate-transformation-examples/pull/10, which looks like

<img width="2149" height="536" alt="image" src="https://github.com/user-attachments/assets/a8292348-6744-41ad-ba3d-b15fe000f3cb" />

on the RHS the "anatomical" coordinate system is like your "rotated" system, it only appears and is transformed to in one of the images.

> and MUST contain axis information pertaining to physical coordinates.

Reviewer 4: In the graph above, the order of coordinate systems in "overview.ome.zarr" (on the right) is what matters. Because "physical" is first in the list, the reference to "overview.ome.zarr" from the top-level collection (parent group) points to "physical" instead of "anatomical".

> and MUST contain axis information pertaining to physical coordinates.

Reviewer 2: That's a nice graph example, thanks.

Do I understand it correctly to say that the `overview/physical` coordinateSystem is the *output* of the `overview.ome.zarr` coordinateSystem?
In that case I can see how that makes sense. I was previously only thinking of child images as "input".

> and MUST contain axis information pertaining to physical coordinates.

Reviewer 4: > Do I understand it correctly to say that the overview/physical coordinateSystem is the output of the overview.ome.zarr coordinateSystem?

I guess conceptually yes, but in the metadata that's not explicitly specified but implicit because "physical" is the first item in the coordinate system list. My way of thinking about it is `overview.ome.zarr` is a OME-Zarr image (not a coordinate system), and when 'input' or 'output' in a coordinate transform is given as the path `overview.ome.zarr`, `ome-zarr-models` resolves that to the first coordinate system listed inside the `overview.ome.zarr` metadata. These resolutions between groups are dashed lines in my graphs.

It should be used for viewing and processing unless a use case dictates otherwise.
It will generally be a representation of the image in its native physical coordinate system.

The following MUST hold for all coordinate systems inside multiscales metadata.
The length of "axes" must be between 2 and 5
and MUST be equal to the dimensionality of the Zarr arrays storing the image data (see "datasets:path").
The "axes" MUST contain 2 or 3 entries of "type:space"
and MAY contain one additional entry of "type:time"
and MAY contain one additional entry of "type:channel" or a null / custom type.
In addition, the entries MUST be ordered by "type" where the "time" axis must come first (if present),
followed by the  "channel" or custom axis (if present) and the axes of type "space".
If there are three spatial axes where two correspond to the image plane ("yx")
and images are stacked along the other (anisotropic) axis ("z"),

> the spatial axes SHOULD be ordered as "zyx".

Reviewer: rfc-3 would also remove this, but the main goal of rfc-3 was to remove restrictions on the number / semantics of axes. The text here is just a recommendation, not a restriction.

whose value is a list of dictionaries that define a transformation
that maps Zarr array coordinates for this resolution level to the "intrinsic" coordinate system
(the last entry of the `coordinateSystems` array).
The transformation is defined according to [transformations metadata](#transformation-types).
The transformation MUST take as input points in the array coordinate system
corresponding to the Zarr array at location `path`.
The value of "input" MUST equal the value of `path`, 
implementations should always treat the value of `input` as if it were equal to the value of `path`.
The value of the transformation’s `output` MUST be the name of the "intrinsic" [coordinate system](#coordinatesystems-metadata).

This transformation MUST be one of the following:

* A single scale or identity transformation
* A sequence transformation containing one scale and one translation transformation.

In these cases, the scale transformation specifies the pixel size in physical units or time duration.
If scaling information is not available or applicable for one of the axes,
the value MUST express the scaling factor between the current resolution
and the first resolution for the given axis,

> defaulting to 1.0 if there is no downsampling along the axis.

Reviewer 4: I agree this section is confusing. Since this section is specifically concerned with transforms in the multiscales (not scale transforms in other parts of the metadata), how about simply dropping the last two sentences and just having

> In these cases, the scale transformation specifies the pixel size in physical units or time duration.

?

If not, and the other sentences are required, then it's not clear 1) in what cases is scaling information not available? 2) in what cases is scaling information not applicable ?3) what is the "first resolution for the given axis"?

This is strongly recommended
so that the the "intrinsic" coordinate system of the image avoids more complex transformations.

If applications require additional transformations,
each `multiscales` dictionary MAY contain the field `coordinateTransformations`,
describing transformations that are applied to all resolution levels in the same manner.
The value of `input` MUST equal the name of the "intrinsic" coordinate system.
The value of `output` MUST be the name of the output coordinate System
which is different from the "intrinsic" coordinate system.

> Each `multiscales` dictionary SHOULD contain the field `name`.

Reviewer: If fixing this is out of scope for rfc5, what mechanism should be used to fix it? Are we waiting for someone with the energy to kick off the rfc process to add a type to this field?

> Each `multiscales` dictionary SHOULD contain the field `name`.

Author: Hey Davis, I'll lay out everything further down (just haven't had the time yet), I hope it will become clearer then. I just marked this as resolved to make sure this can go forward, not to dismiss the debate about it. I just think this review section isn't the right place to fix it.

> Each `multiscales` dictionary SHOULD contain the field `name`.

Reviewer: sounds good!

> Each `multiscales` dictionary SHOULD contain the field `type`,

Author: See [previous](https://github.com/ome/ngff/pull/350/files#r2481043489) comment

which gives the type of downscaling method used to generate the multiscale image pyramid.
It SHOULD contain the field "metadata",
which contains a dictionary with additional information about the downscaling method.

````{admonition} Example

A complete example of json-file for a 5D (TCZYX) multiscales with 3 resolution levels could look like this:

```json
{
  "zarr_format": 3,
  "node_type": "group",
  "attributes": {
    "ome": {
      "version": "0.5",
      "multiscales": [
        {
          "name": "example",
          "coordinateSystems": [
              {
                "name": "intrinsic",
                "axes": [
                  { "name": "t", "type": "time", "unit": "millisecond" },
                  { "name": "c", "type": "channel" },
                  { "name": "z", "type": "space", "unit": "micrometer" },
                  { "name": "y", "type": "space", "unit": "micrometer" },
                  { "name": "x", "type": "space", "unit": "micrometer" }
                ]
              }
          ],
          "datasets": [
            {
              "path": "0",

>               "coordinateTransformations": [

Author: Good question - but since this part of the example hasn't changed with respect to 0.5, I'd place that issue outside the scope of this rfc.

                {
                  // the voxel size for the first scale level (0.5 micrometer)
                  // and the time unit (0.1 milliseconds), which is the same for each scale level
                  "type": "scale",
                  "scale": [1.0, 1.0, 0.5, 0.5, 0.5],
                  "input": "0",
                  "output": "intrinsic"
                }
              ]
            },
            {
              "path": "1",
              "coordinateTransformations": [
                {
                  // the voxel size for the second scale level (downscaled by a factor of 2 -> 1 micrometer)
                  // and the time unit (0.1 milliseconds), which is the same for each scale level
                  "type": "scale",
                  "scale": [1.0, 1.0, 1.0, 1.0, 1.0],
                  "input": "1",
                  "output": "intrinsic"
                }
              ]
            },
            {
              "path": "2",
              "coordinateTransformations": [
                {
                  // the voxel size for the third scale level (downscaled by a factor of 4 -> 2 micrometer)

>                   // and the time unit (0.1 milliseconds), which is the same for each scale level

Author: 7f93e2c783dff1a03f3a6233f200cd09716f58db
